docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,446 @@
+
+![Figure 7](https://i.imgur.com/DsNzh0Y.png)
+
+
+The Cluster API Controller manager receives an event for the updated AWSInfrastructureConfig resource and enqueues a reconciliation request for the associated Machine. The Machine Controller understands that the infrastructure is ready by checking that the AWSInfrastructureConfig.Status.Addresses field has been populated and proceeds to wait for a Kubernetes Node matching the Machine.ProviderID in “Ready” state. After a Node is found, the Machine Controller sets Machine.Status.NodeRef, Machine.Status.Phase to “Ready” and updates the Machine object.
+
+![Figure 8](https://i.imgur.com/K8VUHP4.png)
+
+The Machine has now become a Kubernetes Node and ready to be used.
+
+
+### User Stories
+
+#### As a Kubernetes operator, I’d like to provide custom bootstrap data without the use of a Kubernetes controller.
+
+#### As a Kubernetes operator, I’d like to monitor the progress of fulfilling a Machine and understand what errors, if any, have been reported by the controllers involved.
+
+#### As an infrastructure provider author, I would like to build the fewest number of components possible to support the full cluster-api.
+
+#### As an infrastructure provider author, I would like to take advantage of the kubernetes API to provide validation for provider-specific data needed to provision a machine.
+
+#### As an infrastructure provider consumer, I would like to have validation for the provider-specific data I need to give the system to have it provision a machine.
+
+#### As an infrastructure  provider author, I would like to build a controller to manage provisioning machines using tools of my own choosing.
+
+#### As an infrastructure provider author, I would like to build a controller to manage provisioning machines without being restricted to a CRUD API.
+
+### Implementation Details/Notes/Constraints
+
+#### Machine Controller Role
+The Machine Controller lives within Cluster API manager alongside MachineSet and MachineDeployment controllers. This controller is the only controller having write permissions to Machine objects. The controller responsibility focuses on state transitions and operations including but not limited to:
```

Comment:
I think saying "This controller is the only controller having write permissions to Machine objects." is a little strong. There may be completely legitimate reasons for other controllers to write to Machine objects. We don't want the controllers creating the server behind the Machine to do that as they update status, but I don't think we want to say "nothing other than the machine controller is allowed to write to machines".

Reply:
It would be nice if a developer's first thought is "I shouldn't touch Machines`, but I agree that "must" is probably too strong here.

Reply:
Reworded